argoverse/evaluation/detection_utils.py

```diff
@@ -0,0 +1,201 @@
+# <Copyright 2020, Argo AI, LLC. Released under the MIT license.>
+"""Detection utilities for the Argoverse detection leaderboard.
+
+Accepts detections (in Argoverse ground truth format) and ground truth labels
+for computing evaluation metrics for 3d object detection. We have five different,
+metrics: mAP, ATE, ASE, AOE, and DCS. A true positive for mAP is defined as the
+highest confidence prediction within a specified Euclidean distance threshold
+from a bird's-eye view. We prefer these metrics instead of IoU due to the
+increased interpretability of the error modes in a set of detections.
+
+"""
+
+from enum import Enum, auto
+from typing import List, Tuple
+
+import numpy as np
+import pandas as pd
+from scipy.spatial.distance import cdist
+from scipy.spatial.transform import Rotation as R
+
+from argoverse.data_loading.object_label_record import ObjectLabelRecord
+from argoverse.utils.transform import quat_argo2scipy_vectorized
+
+
+class AffFnType(Enum):
+    CENTER = auto()
+
+
+class DistFnType(Enum):
+    TRANSLATION = auto()
+    SCALE = auto()
+    ORIENTATION = auto()
+
+
+class InterpType(Enum):
+    ALL = auto()
+
+
+class FilterMetric(Enum):
+    EUCLIDEAN = auto()
+
+
+def filter_instances(
+    instances: List[ObjectLabelRecord], target_class_name: str, filter_metric: FilterMetric, max_detection_range: float
+) -> np.ndarray:
+    """Filter the GT annotations based on a set of conditions (class name and distance from egovehicle).
+
+    Args:
+        instances: The instances to be filtered (N, ).
+        target_class_name: The name of the class of interest.
+        filter_metric: The range metric used for filtering.
+        max_detection_range: The maximum distance for range filtering.
+
+    Returns:
+        The filtered annotations.
+    """
+    instances = np.array([instance for instance in instances if instance.label_class == target_class_name])
+
+    if filter_metric == FilterMetric.EUCLIDEAN:
+        centers = np.array([dt.translation for dt in instances])
+        filtered_annos = np.array([])
+
+        if centers.shape[0] > 0:
+            dt_dists = np.linalg.norm(centers, axis=1)
+            filtered_annos = instances[dt_dists < max_detection_range]
+    else:
+        raise NotImplementedError("This filter metric is not implemented!")
+    return filtered_annos
+
+
+def rank(dts: List[ObjectLabelRecord]) -> Tuple[np.ndarray, np.ndarray]:
+    """Get the rankings for the detections, according to detector confidence.
+
+    Args:
+        dts: Detections (N,).
+
+    Returns:
+        ranks: The ranking for the detections (N, ).
+        scores: The detection scores (N, ).
+    """
+    scores = np.array([dt.score for dt in dts])
+    ranks = scores.argsort()[::-1]
+    ranked_detections = dts[ranks]
+    return ranked_detections, scores[:, np.newaxis]
+
+
+def interp(prec: np.ndarray, method: InterpType = InterpType.ALL) -> np.ndarray:
+    """Interpolate the precision over all recall levels.
```

Comment:
it is a bit of a stretch to call this interpolation, since we are using the values at all data points, rather than interpolating at 11 (PASCAL) 40 or 100 points of recall.

I would mention that this function forces monotonicity of precision, from right to left.

Reply:
What's the difference between enforcing right-left monotonicity and interpolated precision/recall? I thought they were the same, likeslide 27 in: https://www.cl.cam.ac.uk/teaching/1415/InfoRtrv/lecture5.pdf

Reply:
Because if it's the difference between using all data points and using fixed recall points, doesn't that contradict the later comment about all-type interpolation?